### PR TITLE
Fixing Example and Adding Keyword Arguments as option for parameters

### PR DESCRIPTION
**Pull request by JonahKr**
1. This fixes a small typo in the async example
2. Just by adding a few lines of code, it enables the functionality to pass parameters in code. (Issue #15 ) e.g:
`python
app = HermesApp("AdviceApp", host = "192.168.178.123", port = 12183)
`

Not sure if it should be restricted to the existing parameters (or some parameters exclusively) or not. I'm leaning towards no restrictions.

### CODE REVIEW

**Comment by koenvervloesem** [comment]
Thanks! An easy addition, so it seems like a no-brainer. I'll have a look at it soon.

**Comment by JonahKr** [comment]
One issue which just came into my mind is that we would need to decide which configuration will override in case of assigning different values as cli arguments and keyword arguments.
Mainly this is to be decided by the user but I see an error potential here.

**Comment by koenvervloesem** [comment]
Ok, I have been playing with this and this looks nice.

So currently the parameters you pass to the `HermesApp` initialization take precedence over the options on the command line. To me it seems like the other way around is more logical, no?

**Comment by JonahKr** [comment]
> 
> 
> Ok, I have been playing with this and this looks nice.
> 
> So currently the parameters you pass to the `HermesApp` initialization take precedence over the options on the command line. To me it seems like the other way around is more logical, no?

Yes thats right... I initially thought it wouldn't matter because the user has to decide which option to use but by ranking cli Arguments higher, we actually add a feature for the App developer to set standard parameters which can be overwritten by the user if his settings are different. I will implement a fix.

**Comment by JonahKr** [comment]
@koenvervloesem i think i fixed the priorities

File: rhasspyhermes_app/__init__.py [hunk 1]
```diff
@@ -106,6 +107,11 @@ def __init__(
         # Parse command-line arguments
         self.args = parser.parse_args()
 
+        # Option to set all parameters as keyword arguments !!! CLI arguments are rated higher priority
+        args_dict = vars(self.args)
+        for key, value in kwargs.items():
+            args_dict[key] = args_dict[key] if key in args_dict else value
```
Comment (maxbachmann) [inline rhasspyhermes_app/__init__.py:113]: I think this should be simply written as:
```python
args_dict = {**kwargs, **vars(self.args)}
```
instead

Comment (JonahKr) [reply rhasspyhermes_app/__init__.py:113]: This is so much more cleaner in every way jeeeeez.... i didn't even know of this feature...
```suggestion
        # Option to set all parameters as keyword arguments !!! CLI arguments are rated higher priority
        args_dict = {**kwargs, **vars(self.args)}
```

**Comment by maxbachmann** [comment]
@JonahKr I missed that you use args_dict only to edit self.args, while my changed version creates a new dict

**Comment by JonahKr** [comment]
@maxbachmann yes me too.... i guess its fixed now

**Comment by maxbachmann** [comment]
Yes should be

**Comment by maxbachmann** [comment]
@JonahKr  Here is a better solution:

```python
self.args = argparse.Namespace(**kwargs, **vars(parser.parse_args()))
```

I think this makes more clear, that it is changing this Namespace object aswell

**Comment by JonahKr** [comment]
> ```python
> args = {**kwargs, **vars(parser.parse_args())}
> self.args = argparse.Namespace(**args)
> ```



> I think this makes more clear, that it is changing this Namespace object aswell, which was really easy to miss in the initial solution

my solution did indead change the Namespace object. But the CLI-arguments have standard values which therefore causes them to not get overwritten...

**Comment by JonahKr** [comment]
maybe we should finish this thread afterwards, close the pull request and create a new one, or otherwise a ton of commits without content thrash up the history

**Comment by maxbachmann** [comment]
> But the CLI-arguments have standard values which therefore causes them to not get overwritten...

thats true (this should have already the same problem in your initial priority change)

**Comment by JonahKr** [comment]
> (this should have already the same problem in your initial priority change

yes but i didn't have the environment to test it and tried to replicate the environment. but now i finished the setup off a rhasspy container on my laptop so i tried it

**Comment by koenvervloesem** [comment]
Ok I haven't tried to fix this code yet, but because it seems more difficult than I initially thought, I have written some test cases, so we can see for sure which corner cases are failing while trying to fix this. I think I have covered all cases in the test, feel free to add tests if you think I missed something. The default CLI arguments are indeed problematic:
```

mocker = <pytest_mock.plugin.MockFixture object at 0x7f382d83aa60>

    def test_arguments_in_init(mocker):
        """Test whether arguments are set up correctly while initializing a HermesApp object."""
        app = HermesApp(
            "Test arguments in init",
            mqtt_client=mocker.MagicMock(),
            host="rhasspy.home",
            port=8883,
            tls=True,
            username="rhasspy-hermes-app",
            password="test",
        )
    
>       assert app.args.host == "rhasspy.home"
E       AssertionError: assert 'localhost' == 'rhasspy.home'
E         - rhasspy.home
E         + localhost

tests/test_arguments.py:59: AssertionError
```

**Comment by koenvervloesem** [comment]
Ok I have something that seems to work, but it's a bit ugly :-) `make test` passes, but I have to look at the failing `make check`.

Can you review this?

**Comment by maxbachmann** [comment]
Wouldn't this break when someone passes a default argument value on purpose to the cli?

**Comment by koenvervloesem** [comment]
You're right, I even thought about it this morning, but forgot this when working on the fix. I have covered this test case now in the tests. Still thinking about the best way to fix this...

**Comment by maxbachmann** [comment]
yes I think it would be totally confusing when some command-line arguments take precedence while others do not.

**Comment by koenvervloesem** [comment]
So, apparently [argparse](https://docs.python.org/3/library/argparse.html#module-argparse) can't distinguish between an argument that it assigns its default value because you don't supply it and the same argument that you assign its default value explicitly. At least I haven't found a way. Maybe one of you has more luck finding it?

**Comment by maxbachmann** [comment]
@koenvervloesem could you add another test for the case that the user provides his own parser, that might both include optional and required arguments, but should apparently follow the same system that cli arguments take precendence over the kwargs

File: rhasspyhermes_app/__init__.py [hunk 2]
```diff
@@ -103,8 +105,18 @@ def __init__(
         # Add default arguments
         hermes_cli.add_hermes_args(parser)
 
-        # Parse command-line arguments
-        self.args = parser.parse_args()
+        # overwrite argument defaults inside parser with argparse.SUPPRESS
+        # so arguments that are not provided get ignored
+        suppress_parser = deepcopy(parser)
+        for action in suppress_parser._actions:
+            action.default = argparse.SUPPRESS
```
Comment (maxbachmann) [inline rhasspyhermes_app/__init__.py:112]: it works now even though depending on the internals this way is probably not the cleanest solution :joy:

**Comment by koenvervloesem** [comment]
Thanks, @maxbachmann! I added another test for verifying that it still works when you supply your own argument parser.

If @JonahKr can have a look at the result, I think this is ready to be merged and create a new release :-)

**Review by JonahKr — COMMENTED** [review]
I cannot test it atm but the code looks good. If both of you tested it and it works, it's ready to merge in my opinion.

**Comment by JonahKr** [comment]
But maybe merge it as 1 commit.😂 Because this hellscape of wrong code should never touch the sight of light again.

**Comment by maxbachmann** [comment]
> But maybe merge it as 1 commit.joy Because this hellscape of wrong code should never see the light of day again...
I agree this should probably simply be squashed when merging
